src/mono/mono/mini/mini-runtime.c

```diff
@@ -5395,3 +5401,33 @@ mini_alloc_jinfo (MonoJitMemoryManager *jit_mm, int size)
 		return (MonoJitInfo*)mono_mem_manager_alloc0 (jit_mm->mem_manager, size);
 	}
 }
+
+void
+mono_set_runtime_init_callback (MonoRuntimeInitCallback callback)
+{
+	runtime_init_callback = callback;
+}
+
+/*
+* Default implementation invoking runtime init callback in native-to-managed wrapper.
+*/
+void
+mono_invoke_runtime_init_callback (void)
+{
+	if (G_UNLIKELY (runtime_init_callback)) {
+		guint64 thread_id = mono_native_thread_os_id_get ();
+		if (runtime_init_callback && (runtime_init_thread_id == thread_id))
+			return;
+
+		while (mono_atomic_cas_i64 ((volatile gint64 *)&runtime_init_thread_id, (gint64)thread_id, (gint64)G_MAXUINT64) != (gint64)G_MAXUINT64)
+			g_usleep (1000);
+
+		if (runtime_init_callback) {
+			if (!mono_thread_info_current_unchecked ())
+				runtime_init_callback ();
+			runtime_init_callback = NULL;
```

Comment:
Hmm... or maybe not? I guess our `mono_atomic_xchg_i64` is meant to be a full read/write barrier.

Reply:
I mainly looked at mono_atomic_xchg_i64 for the write barrier, but we should probably prevent reorder here as well.

Reply:
Yeah, but if we should be real strict on the acquire/release semantics you should have the barrier between call to runtime_init_callback and setting runtime_init_callback = NULL and not after.

Reply:
What could happen is another thread sees runtime_init_callback and returns, before it has seen all side effects of runtime_init_callback call.

Reply:
I hardened the implementation using acquire/release semantics on the callback + atomic ops on the thread id to prevent potential re-ordering between resetting thread id and callback. Since the acquire semantics happens on the hot path, it will have impact on platforms that needs additional barriers. Sometimes you drop the acquire barrier, but in this case, where we depend on init of GOT slots right after return from the callback, I believe it will cause issues sooner rather than later on platforms with weak memory models. If it turns out to be a performance bottleneck, there is always the possibility to implement a custom callback that could take other decisions around thread safety.